build: a number of fixes on Makefile and nix-shell

tiagolobocastro opened the pull request:
**What this PR does?**:
Ensure the namespace is always set correctly to openebs by default.
Set the image pull policy properly (wrong var was used...)

The non-buildx Dockerfile copies a binary built outside of the docker
build, which may be broken if the dependencies are not met.
This happens on nix-shell as the deps exist on /nix/store.
Get around this by building static binaries in the nix-shell.

Pin a few packages version which are not compatible on latest.
Perhaps all of them should be pinned!
Remove a few uses of sudo which should not be required?
Add a few packages missing from nix-shell
Allow make clean to work properly
Don't bootstrap by default, though this may be discussed further.
Setup goenv env variables on github actions workflows
Add gitignore for the coverage data

**Does this PR require any upgrade changes?**:

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Any additional information for your reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

tiagolobocastro committed:
build: a number of fixes on Makefile and nix-shell

Pin a few packages version which are not compatible on latest.
Perhaps all of them should be pinned!
Remove a few uses of sudo which should not be required?
Add a few packages missing from nix-shell
Allow make clean to work properly
Don't bootstrap by default, though this may be discussed further.
Setup goenv env variables on github actions workflows
Add gitignore for the coverage data

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented on shell.nix:
```diff
@@ -16,14 +16,16 @@ pkgs.mkShell {
     export TMPDIR=$(pwd)/nix/.tmp
     export PATH=$GOPATH/bin:$PATH
     mkdir -p "$TMPDIR"
-    make bootstrap
```
What do you think @mhkarimi1383? Should we keep this automatically?

mhkarimi1383 replied:
shell.nix file made to make a development environment automatically for Nix users, So I think we should keep this to run automatically it will do nothing and just install requirements like other requirements that we are installing in the same file

mhkarimi1383 replied:
shell.nix file made to make a development environment automatically for Nix users, So I think we should keep this to run automatically it will do nothing and just install requirements like other requirements that we are installing in the same file

tiagolobocastro replied:
A valid point, reverted.

tiagolobocastro commented:
Ah looks like CI failed because GOPATH is not set or empty. Is this expected?
CC @dsharma-dc @Abhinandan-Purkait

tiagolobocastro commented:
I'm also a bit confused with the `bdd-tests` on PR, it seems like it's testing against develop rather than the code built in this PR or am I'm missing something?

mhkarimi1383 commented on shell.nix:
```diff
@@ -16,14 +16,16 @@ pkgs.mkShell {
+    util-linux
+  ] ++ pkgs.lib.optional (builtins.getEnv "IN_NIX_SHELL" == "pure") docker;
   shellHook = ''
-    export HOME=${builtins.getEnv "HOME"}
```
HOME env variable should be inherited explicitly when we are using this in GitHub Actions

mhkarimi1383 replied:
If GitHub Actions are working fine we can ignore this :)

mhkarimi1383 commented:
> Ah looks like CI failed because GOPATH is not set or empty. Is this expected?
> CC @dsharma-dc @Abhinandan-Purkait 

I think that's because HOME env inheritance is removed

mhkarimi1383 commented:
Can you make builds static?
Because docker builds are not working in Nix and some OSes (Because of shared C libs)

See
https://github.com/openebs/lvm-localpv/pull/335#issuecomment-2564656408

Abhinandan-Purkait commented on Makefile:
```diff
@@ -222,15 +226,15 @@ lvm-driver-image: lvm-driver
 
 .PHONY: ansible-runner-image
 ansible-runner-image:
 	@echo "------------------"
 	@echo "--> Build ansible-runner image for lvm-localpv e2e-tests" 
 	@echo "------------------"
-	sudo docker build . -f e2e-tests/Dockerfile -t ${IMAGE_ORG}/lvm-localpv-e2e:ci
+	docker build . -f e2e-tests/Dockerfile -t ${IMAGE_ORG}/lvm-localpv-e2e:ci
```
Why do we need the ansible-runner-image? We have migrated the e2e right?

tiagolobocastro replied:
Not sure, that was already there...
I can remove it?

Abhinandan-Purkait commented:
> Ah looks like CI failed because GOPATH is not set or empty. Is this expected? CC @dsharma-dc @Abhinandan-Purkait

The setup go action step sets it to home/runner/go

Abhinandan-Purkait commented:
> I'm also a bit confused with the `bdd-tests` on PR, it seems like it's testing against develop rather than the code built in this PR or am I'm missing something?

Which bit are you talking about? It checkout the PR code and build image for it and then uses it from the built image locally?

tiagolobocastro commented:
> > Ah looks like CI failed because GOPATH is not set or empty. Is this expected?
> > CC @dsharma-dc @Abhinandan-Purkait
> 
> I think that's because HOME env inheritance is removed

No: https://github.com/openebs/lvm-localpv/actions/runs/12331398437/job/34417991555#step:7:34

Ah I wanted to ask you, why is this needed? In my case it's always there, even without this line.

tiagolobocastro commented:
> > Ah looks like CI failed because GOPATH is not set or empty. Is this expected? CC @dsharma-dc @Abhinandan-Purkait
> 
> The setup go action step sets it to home/runner/go

Fixed by taking go env, PTAL

tiagolobocastro commented:
> > I'm also a bit confused with the `bdd-tests` on PR, it seems like it's testing against develop rather than the code built in this PR or am I'm missing something?
> 
> Which bit are you talking about? It checkout the PR code and build image for it and then uses it from the built image locally?

Discussed offline, I will update here...
So on GHA minikube is setup with driver none, and thus the images are picked up from the host this way.. this is confusing and should be documented.
Also this makes local testing on custom cluster not very straightforward with current script

tiagolobocastro commented:
> Can you make builds static? Because docker builds are not working in Nix and some OSes (Because of shared C libs)
> 
> See [#335 (comment)](https://github.com/openebs/lvm-localpv/pull/335#issuecomment-2564656408)

The problem here is that the dockerfile is simply copying the binary to the container. On non-nix binaries this might work though it might also fail at runtime potentially if the OS libs are not compatible with the container libs.

On the other hand the buildx is building inside the docker build, which is a bit better, though it seems that the same problem may arise from what I can tell looking at it? 

I'll make use of musl-gcc on the nix-shell and make build static just for the nix-shell for now? Or shall we make it static for docker buildx as well? CC @Abhinandan-Purkait @dsharma-dc

tiagolobocastro committed:
build(nix-shell): use musl libc and static build

The non-buildx Dockerfile copies a binary built outside of the docker
build, which may be broken if the dependencies are not met.
This happens on nix-shell as the deps exist on /nix/store.
Get around this by building static binaries in the nix-shell.
We now depend on golang-alpine image and install musl-dev as well.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

mhkarimi1383 commented:
@tiagolobocastro 

Yes, I think RHEL Based users may face the same problem, NixOS is a big example since it's so different that other distributions in terms of paths, etc.

I think we can also change our flow to build directly through the Dockerfile

(For example if we are running our runtime in debian we can also use golang debian based image)

mhkarimi1383 commented:
With static builds we could also publish binary builds later on (I know the usage for that is rare, but we can do it too)

tiagolobocastro committed:
test: default namespace and set image pull policy

Ensure the namespace is always set correctly to openebs by default.
Set the image pull policy properly (wrong var was used...)

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
refactor: deal with shellcheck

A lot of issues on the existing code but seems they are only
warned if the file is modified...

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
A ton of shellcheck warnings were given because I modified the shell file, this is not great to be honest, now it's messed up this PR with a ton of comments.
It would be better to get these in the action, not as comments... -.-

dsharma-dc commented:
> > Can you make builds static? Because docker builds are not working in Nix and some OSes (Because of shared C libs)
> > See [#335 (comment)](https://github.com/openebs/lvm-localpv/pull/335#issuecomment-2564656408)
> 
> The problem here is that the dockerfile is simply copying the binary to the container. On non-nix binaries this might work though it might also fail at runtime potentially if the OS libs are not compatible with the container libs.
> 
> On the other hand the buildx is building inside the docker build, which is a bit better, though it seems that the same problem may arise from what I can tell looking at it?
> 
> I'll make use of musl-gcc on the nix-shell and make build static just for the nix-shell for now? Or shall we make it static for docker buildx as well? CC @Abhinandan-Purkait @dsharma-dc

If we want to make static for nix as well as docker build, then better do that across all localpv repos if they are also following same?

Abhinandan-Purkait commented:
> > Can you make builds static? Because docker builds are not working in Nix and some OSes (Because of shared C libs)
> > See [#335 (comment)](https://github.com/openebs/lvm-localpv/pull/335#issuecomment-2564656408)
> 
> The problem here is that the dockerfile is simply copying the binary to the container. On non-nix binaries this might work though it might also fail at runtime potentially if the OS libs are not compatible with the container libs.
> 
> On the other hand the buildx is building inside the docker build, which is a bit better, though it seems that the same problem may arise from what I can tell looking at it?
> 
> I'll make use of musl-gcc on the nix-shell and make build static just for the nix-shell for now? Or shall we make it static for docker buildx as well? CC @Abhinandan-Purkait @dsharma-dc

Yeah, I think that's a better way in general. If we can do it across all repos.

tiagolobocastro commented:
@Abhinandan-Purkait @dsharma-dc @mhkarimi1383 

Ok, it is done, please give this around review

tiagolobocastro committed:
test: cleanup ginkgo loop vgs

gingko lvm_utils creates some loop devices and vgs, which are not
cleanup in some cases (eg: test failures).
furthermore, even between test cases, leftovers are causing subsequent
tests to fails, but that's another matter that needs fixing.
also ensure make clean cleans these up

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
build(nix-shell): add missing lvm packages to pure nix-shell

Adds lvm2 and jq
Also ensure sudo works in a pure shell

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

mhkarimi1383 commented:
I have tested the build locally and it was working fine
